components/submit-expense/useExpenseForm.ts

```diff
@@ -0,0 +1,399 @@
+        hasPayeeSlug: !!values.payeeSlug,
+      },
+    });
+
+    const account = query.data?.account;
+    const host = account && 'host' in account ? account.host : null;
+
+    if (host) {
+      options.expensePolicy = host.expensePolicy;
+      options.expenseTags = host.expensesTags;
+      options.isAccountingCategoryRequired = userMustSetAccountingCategory(loggedInUser, account, host);
+      options.accountingCategories = host.accountingCategories.nodes;
+
+      if (options.isAccountingCategoryRequired) {
+        options.schema = options.schema.extend({
+          accountingCategoryId: z.string(),
+        });
+      }
+    }
+
+    if ((query.data?.account?.supportedExpenseTypes ?? []).length > 0) {
+      options.schema = options.schema.extend({
+        expenseTypeOption: z
+          .nativeEnum(ExpenseTypeOption)
+          .refine(v => query.data.account.supportedExpenseTypes.includes(expenseTypeFromOption(v)), {
+            message: 'expense type not supported',
+          }),
+      });
+      options.supportedExpenseTypes = query.data.account.supportedExpenseTypes;
+    }
+
+    if (query.data?.loggedInAccount) {
+      options.payoutProfiles = getPayoutProfiles(query.data.loggedInAccount);
+    }
+
+    options.supportedCurrencies = Object.values(Currency);
+    if (query.data?.payee && values.payoutMethodId && values.expenseTypeOption) {
+      const supportedCurrencies = getSupportedCurrencies(query.data?.account, {
+        payee: query.data?.payee,
+        payoutMethod: query.data?.payee?.payoutMethods?.find(p => p.id === values.payoutMethodId),
+        type: expenseTypeFromOption(values.expenseTypeOption),
+        currency: account.currency,
+      });
+
+      options.schema = options.schema.extend({
+        expenseCurrency: z.string().refine(v => supportedCurrencies.includes(v), {
+          message: `Currency must be one of: ${supportedCurrencies.join(',')}`,
+        }),
+      });
+
+      options.supportedCurrencies = supportedCurrencies;
+    }
+
+    options.schema = options.schema.extend({
+      expenseItems: z.array(
+        z.object({
+          description: z.string().min(1),
+          url:
+            expenseTypeFromOption(values.expenseTypeOption) === ExpenseType.RECEIPT
+              ? z.string().url()
+              : z.string().url().optional(),
+          date: z.string(),
+          amount: z.object({
+            valueInCents: z.number(),
+            currency: z.string().refine(v => Object.values(Currency).includes(v as Currency), {
+              message: `Currency must be one of: ${Object.values(Currency).join(',')}`,
+            }),
+          }),
+        }),
+      ),
+    });
+
+    options.allowExpenseItemAttachment = options.allowExpenseItemCurrencyChange =
+      expenseTypeFromOption(values.expenseTypeOption) === ExpenseType.RECEIPT;
+
+    return options;
+  } catch (err) {
+    return options;
+  }
+}
+
+export function useExpenseForm(opts: {
```

Comment:
@Betree using this hook to take most, if not all, logic from the rendering of form fields. Its a huge function and still missing the additional logic existing in the current form, but hopefully it will keep everything contained here.

Reply:
Good call. I considered doing something like this the last time I looked at this form; the number of `useEffects` has increased a lot since the original implementation and keeping everything in the component is not great for maintainability.